api/v1alpha1/operator_types.go

```diff
@@ -27,8 +27,14 @@ type OperatorSpec struct {
 	PackageName string `json:"packageName"`
 }
 
+const (
+	TypeReady = "Ready"
+)
+
 // OperatorStatus defines the observed state of Operator
-type OperatorStatus struct{}
+type OperatorStatus struct {
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
```

Comment:
I haven't seen (and I haven't looked very hard) patchStrategy and patchMergeKey on conditions before. Is this something suggested in official k8s docs somewhere?

I'm wondering what the implications are? It seems like this would permit two writers to server-side apply two different types concurrently? Do we want that?

Reply:
Yes, these are the recommended attributes:
https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition
What it should mean is that you can only have a single instance of each type.

Reply:
Do we also need the other tags for `Conditions` that are listed in the example `FooStatus` on that page? I'm not actually sure, but it's definitely worth adding them and running `make generate manifests` to see what they do to the generated code and/or CRD file.
```yaml
type FooStatus struct{
    // Represents the observations of a foo's current state.
    // Known .status.conditions.type are: "Available", "Progressing", and "Degraded"
    // +patchMergeKey=type
    // +patchStrategy=merge
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`

    // other fields
}
```

Reply:
It does add something (`x-kubernetes-list`) values. So I might as well.

Reply:
> you can only have a single instance of each type.

Another follow-up/TODO/ground rule/invariant:

For every `condition.Type` that we define, that type has a condition set no matter what code path is taken through the reconciler. Every single time reconcile is called, we make sure to evaluate what the value of each `condition.Type` should be and we make sure it is set that way.

TL;DR:
1. Reconcile is idempotent. Same input always results in the same output.
2. Output always includes exactly one of each condition type.

Where input is defined as: "Operator CR (minus its status) + existing cluster state"
And output is defined as: "Operator CR status + new desired cluster state"